Rectangular maps, maps generator, bugfixes (part 1)

r3versi opened the pull request:
This is part 1 of maps related work I've done together with Cpt. Klutz during these days. 

- **Rectangular maps support**: `TileMap` rectangular constructor and `MapShape` class added
- UI change to let the user pick rectangular maps
- `Perlin` noise refactor:
  - gradients bugfix
  - in order to have smooth and coherent multitiles structure the value of perlin noise for each tile is calculated based on `WorldCoords` rather than `HexCoords` (to reflect the actual placing of tiles on screen)
  - utility function to combine several octaves
- `MapGenerator` refactor:
  - equal `mapParameters` generate identical maps: all `RNG` calls a common `RNG` properly seeded
  - the landmass generator is a singleton
  - parametrized several `mapParameters` to let the user change the default value in future
- MapTypes: `Pangea`, `Continents` and `Perlin` landmass generator changed to perlin, the only type using the cellular automata landmass generator is `Default`. 

- `NewGameScreen` & `MapParametersTable` UI changes: all labels aligned to left.
- Bugfixes on `NewGameScreen`:
  - the `MapType` cannot be set to `Generated` anymore (as `Generated` is a _macrotype_ rather than a _type_)
  - generating a non-valid map doesn't force the user to close and reopen the app to try creating a new game.

Close #1219

yairm210 commented on core/src/com/unciv/logic/HexMath.kt:
```diff
@@ -59,6 +85,20 @@ object HexMath {
+    fun cubic2EvenQCoords(cubicCoord: Vector3): Vector2 {
+        return Vector2(cubicCoord.x, cubicCoord.z + (cubicCoord.x + (cubicCoord.x.toInt() and 1)) / 2)
+    }
+    fun evenQ2CubicCoords(evenQCoord: Vector2): Vector3 {
```
Beg pardon, but what are evenQ coordinates?

r3versi replied:
Sorry for not clearing this. It's a system of coordinates for rectangular hex maps where the hexagon coordinates are assigned as in figure below:
![evenq](https://user-images.githubusercontent.com/17114100/73777895-8aebf580-478a-11ea-9658-5a3f8d985a50.png)

In creating a rectangular map i think this coordinate system feels natural, so i generate the map using this coordinate system and then convert it to our Hex coordinate system.

yairm210 commented on core/src/com/unciv/ui/newgamescreen/NewGameScreenOptionsTable.kt:
```diff
@@ -36,55 +44,47 @@ class NewGameScreenOptionsTable(val newGameScreen: NewGameScreen, val updatePlay
+        val mapFileSelectBox = getMapFileSelectBox()
+        savedMapOptionsTable.defaults().pad(5f)
+        savedMapOptionsTable.add("{Map file}:".toLabel()).left()
+        // because SOME people gotta give the hugest names to their maps
```
I thought this was your comment and I lol'd, and then I noticed this is one of mine 0_0

r3versi replied:
_SOME_ comments shall stay forever. :rofl:

yairm210 commented:
Looks really good overall!
I see there was a lot of UI work done, so I'd like a screenshot =)

r3versi commented:
Added the advanced settings tab (to play with specific numbers relative to map generation algorithm) and fixed a bug in which the mapType wasn't correctly updated sometimes.

Here is how the UI looks like now:
![UI1](https://user-images.githubusercontent.com/17114100/73781388-73b00680-4790-11ea-97de-ba9e881f10e3.png)
![UI2](https://user-images.githubusercontent.com/17114100/73781386-73b00680-4790-11ea-8b90-26249ad973a6.png)

yairm210 commented:
"hide advanced settings" maybe?

r3versi commented:
![ui3](https://user-images.githubusercontent.com/17114100/73786620-f12c4480-4799-11ea-95cd-74ebf2b7418f.png)

yairm210 commented:
I would like to ask about:
A. Why are there 2 scales one for land percentage and one for water? Shouldn't they be on the same scale? As someone asked on discord, if together they are less than 100%, then what is the rest?
B. Do you think people will want to return to the defaults?

r3versi commented:
A. As I partially said on discord, we are planning to change other parts of the map generator to try to have more realistic maps that uses quantity like temperature, humidity, biosphere to decide what tiles/feature to spawn. The current parameter names (and values) might seem to be meaningless because of how they affect the map generator. For example, land percent is used only by the Cellular Automata generator to seed the initial map, while water percent is used to spawn extra water zones _after_ the landmass (formation of islands, continents etc.) is done. I think that atm it isn't worth to invest a lot of time in finding suitable names for parameters that we can work to improve and refactor. 

B. Definetely! Will add a `Restore defaults` button.
